ml_mr/estimation/quantile_iv.py

```diff
@@ -290,6 +292,171 @@ def from_results(cls, dir_name: str) -> "QuantileIVEstimator":
+            self,
+            x: torch.Tensor,
+            covars: Optional[torch.Tensor] = None,
+            low_memory: bool = True,
+            alpha: float = 0.1
+    ) -> torch.Tensor:
+        return super().avg_iv_reg_function(x, covars, low_memory, alpha)
+
+    def init_PCA(self, n_components: float = 0.999, random_state: int = 42):
+        pca = PCA(n_components=n_components, random_state=random_state)
+        return pca
+
+    def read_ivdataset(self, dataset: IVDataset) -> Tuple:
+        dl = FullBatchDataLoader(dataset)
+        x, y, ivs, covars = next(iter(dl))
+        return x, y, ivs, covars
+
+    def linear_inference(
+        self,
+        dataset: IVDataset
+    ) -> Callable[[torch.Tensor, Optional[torch.Tensor]], torch.Tensor]:
+        x, y, ivs, covars = self.read_ivdataset(dataset)
+
+        x_hats = self.exposure_network(_cat(ivs, covars))
+        n_quantiles = x_hats.size(1)
+        n = x.size(0)
+
+        eta_bar = None
+        # Loop through all quantiles
+        # Equivalent to the 1/K sum step in equation (6)
+        for j in range(n_quantiles):
+            current_representation = self.outcome_network.get_representation(
+                _cat(x_hats[:, [j]], covars)
+            ) / (n_quantiles)
+
+            if eta_bar is None:
+                eta_bar = current_representation
+            else:
+                eta_bar += current_representation
+
+        assert eta_bar is not None
+
+        # Fit PCA
+        self.pca.fit(eta_bar.detach().numpy())
+        eta_bar_pca = torch.from_numpy(
+            self.pca.transform(eta_bar.detach().numpy())
+        )
+        # Add intercept column
+        eta_bar_pca = torch.hstack((torch.ones((n, 1)), eta_bar_pca))
+
+        # Get the representation of the exposure x 
+        H = self.outcome_network.get_representation(
+            _cat(x, covars)
+        )
+        # Apply PCA to the representation of the exposure x
+        H_pca = torch.from_numpy(
+            self.pca.transform(H.detach().numpy())
+        )
+
+        # Add intercept column
+        H_pca = torch.hstack((torch.ones((n, 1)), H_pca))
+
+        # We now compute the IV regression
+        beta_IV_hat = torch.linalg.inv(
+            eta_bar_pca.T @ H_pca) @ eta_bar_pca.T @ y
+
+        # We compute the variance of the coefficients beta_IV_hat
+        mat_etabar_H_inv = torch.linalg.inv(eta_bar_pca.T @ H_pca)
+        mat_etabar = eta_bar_pca
+        mat_diag = torch.diag(((H_pca @ beta_IV_hat - y) ** 2).reshape(-1))
+        var_beta_IV_hat = (
+            mat_etabar_H_inv
+            @ mat_etabar.T
+            @ mat_diag
+            @ mat_etabar
+            @ mat_etabar_H_inv
+        )
+
+        # Estimated h_hat function
+        def iv_reg(x, covars):
+            eta = self.outcome_network.get_representation(
+                _cat(x, covars)
+            )
+            eta_pca = torch.from_numpy(
+                self.pca.transform(eta.detach().numpy())
```

Comment:
Eventually, if we want to make the implementation cleaner, we could use torch.svd instead of deferring (and depending on) scikit-learn for the PCA.